deps/rabbitmq_mqtt/src/rabbit_mqtt_retained_msg_store_dets.erl

```diff
@@ -2,68 +2,308 @@
+                 match_pattern_words(RestWords, ChildId, State, EdgeAcc)
+              end,
+              Acc,
+              Edges);
+match_pattern_words([<<"#">> | _], NodeId, State, Acc) ->
+  % # matches zero or more words
+  collect_descendants(NodeId, State, [NodeId | Acc]);
+match_pattern_words([Word | RestWords], NodeId, State, Acc) ->
+  case find_edge(NodeId, Word, State) of
+    {ok, ChildId} ->
+      match_pattern_words(RestWords, ChildId, State, Acc);
+    error ->
+      Acc
+  end.
+
+collect_descendants(NodeId, State, Acc) ->
+  Edges = get_all_edges(NodeId, State),
+  lists:foldl(fun({_Key, ChildId}, EdgeAcc) ->
+                 collect_descendants(ChildId, State, [ChildId | EdgeAcc])
+              end,
+              Acc,
+              Edges).
+
+find_edge(NodeId, Word, State) ->
+  Key = {NodeId, Word},
+  case dets:lookup(State#store_state.edge_table, Key) of
+    [{_Key, ToNode}] ->
+      {ok, ToNode};
+    [] ->
+      error
+  end.
+
+get_all_edges(NodeId, State) ->
+  Pattern = {{NodeId, '_'}, '_'},
+  dets:match_object(State#store_state.edge_table, Pattern).
+
+add_edge(FromId, Word, ToId, State) ->
+  Key = {FromId, Word},
+  EdgeEntry = {Key, ToId},
+  ok = dets:insert(State#store_state.edge_table, EdgeEntry),
+  NodeEntry = {ToId, 0, false},
+  ok = dets:insert(State#store_state.node_table, NodeEntry),
+  update_edge_count(FromId, +1, State).
+
+update_edge_count(NodeId, Delta, State) ->
+  case dets:lookup(State#store_state.node_table, NodeId) of
+    [{NodeId, EdgeCount, IsTopic}] ->
+      NewCount = EdgeCount + Delta,
+      ok = dets:insert(State#store_state.node_table, {NodeId, NewCount, IsTopic});
+    [] ->
+      error
+  end.
+
+update_node(NodeId, IsTopic, State) ->
+  case dets:lookup(State#store_state.node_table, NodeId) of
+    [{NodeId, EdgeCount, _OldIsTopic}] ->
+      ok = dets:insert(State#store_state.node_table, {NodeId, EdgeCount, IsTopic});
+    [] ->
+      error
+  end.
+
+maybe_clean_path(NodeId, State) ->
+  case dets:lookup(State#store_state.node_table, NodeId) of
+    [{NodeId, 0, false}] ->
+      Pattern = {'_', '_'},
+      Edges = dets:match_object(State#store_state.edge_table, {Pattern, NodeId}),
+      case Edges of
+        [{{ParentId, Word}, NodeId}] ->
+          remove_edge(ParentId, Word, State),
+          ok = dets:delete(State#store_state.node_table, NodeId),
+          maybe_clean_path(ParentId, State);
+        [] ->
+          ok
+      end;
+    _ ->
+      ok
+  end.
+
+remove_edge(FromId, Word, State) ->
+  ok = dets:delete(State#store_state.edge_table, {FromId, Word}),
+  update_edge_count(FromId, -1, State).
+
+%% Setup functions
+
+open_table(Dir, VHost, Type, TableType) ->
+  Tab = get_table_name(VHost, Type),
+  % Path = get_table_path(Dir, VHost, Type),
+  AutoSave =
+    rabbit_misc:get_env(rabbit_mqtt, retained_message_store_dets_sync_interval, 2000),
+  dets:open_file(Tab,
+                 [{type, TableType},
+                  % {file, Path},
```

Comment:
I still need to figure out why the Path was not accepted and led dets.erl to throw an error.